lib/auth/webauthn/messages_test.go

```diff
@@ -56,3 +57,270 @@ func TestCredentialAssertionResponse_json(t *testing.T) {
+			} else {
+				require.NoError(t, err, "Validate returned err %v, want nil", err)
+			}
+		})
+	}
+}
+
+func TestRequireResidentKey(t *testing.T) {
+	tests := []struct {
+		name    string
+		in      protocol.AuthenticatorSelection
+		want    bool
+		wantErr string
+	}{
+		{
+			name: "nothing set",
+			in:   protocol.AuthenticatorSelection{},
+			want: false,
+		},
+		{
+			name: "mismatch of ResidentKey and RequireResidentKey - return error",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementDiscouraged,
+				RequireResidentKey: protocol.ResidentKeyRequired(),
+			},
+			wantErr: "invalid combinantion of ResidentKey: discouraged and RequireResidentKey: true",
+		},
+		{
+			name: "other mismatch of ResidentKey and RequireResidentKey - return error",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementRequired,
+				RequireResidentKey: protocol.ResidentKeyUnrequired(),
+			},
+			wantErr: "invalid combinantion of ResidentKey: required and RequireResidentKey: false",
+		},
+		{
+			name: "support nil RequireResidentKey",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        "",
+				RequireResidentKey: nil,
+			},
+			want: false,
+		},
+		{
+			name: "ResidentKey preferred result in false",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementPreferred,
+				RequireResidentKey: nil,
+			},
+			want: false,
+		},
+		{
+			name: "ResidentKey required",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey: protocol.ResidentKeyRequirementRequired,
+			},
+			want: true,
+		},
+		{
+			name: "ResidentKey discouraged",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey: protocol.ResidentKeyRequirementDiscouraged,
+			},
+			want: false,
+		},
+		{
+			name: "use RequireResidentKey required if ResidentKey empty",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        "",
+				RequireResidentKey: protocol.ResidentKeyRequired(),
+			},
+			want: true,
+		},
+		{
+			name: "use RequireResidentKey unrequired if ResidentKey empty",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        "",
+				RequireResidentKey: protocol.ResidentKeyUnrequired(),
+			},
+			want: false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			req := &wanlib.CredentialCreation{
+				Response: protocol.PublicKeyCredentialCreationOptions{
+					AuthenticatorSelection: test.in,
+				},
+			}
+			got, err := req.RequireResidentKey()
+			if test.wantErr != "" {
+				require.ErrorContains(t, err, test.wantErr, "RequireResidentKey() returned err = %q, want %q", err, test.wantErr)
+			} else {
+				require.NoError(t, err, "RequireResidentKey() returned err %v, want nil", err)
+				if got != test.want {
+					t.Errorf("RequireResidentKey() = %v, want %v", got, test.want)
+				}
```

Comment:
Since you introduced testify, might as well go for an assert.Equal here.

```suggestion
				assert.Equal(t, test.want, got, "RequireResidentKey mismatch")
```